Git-commit: e9ab28a596dfbe322645389c8b5bdfc0f9fea08e
From: Xie XiuQi <xiexiuqi@huawei.com>
Date: Fri, 26 Feb 2021 20:21:35 +0800
Subject: [PATCH] arm64/mpam: fix a missing unlock in error branch
Patch-mainline: Queued in openEuler repo, version 5.10.0-4.5.0
References: openEuler-21.03
Git-repo: https://gitee.com/openeuler/kernel.git

hulk inclusion
category: bugfix
bugzilla: 48265
CVE: NA

--------------------------------

In error path, we should put the resource and unlock of->kn.

Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Wang ShaoBo <bobo.shaobowang@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Kai Liu <kai.liu@suse.com>
---
 arch/arm64/kernel/mpam_ctrlmon.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/arch/arm64/kernel/mpam_ctrlmon.c b/arch/arm64/kernel/mpam_ctrlmon.c
index ae6afc90a256..02623c1c23e8 100644
--- a/arch/arm64/kernel/mpam_ctrlmon.c
+++ b/arch/arm64/kernel/mpam_ctrlmon.c
@@ -354,8 +354,8 @@ int resctrl_group_mondata_show(struct seq_file *m, void *arg)
 
 	rdtgrp = resctrl_group_kn_lock_live(of->kn);
 	if (!rdtgrp) {
-		resctrl_group_kn_unlock(of->kn);
-		return -ENOENT;
+		ret = -ENOENT;
+		goto out;
 	}
 
 	md.priv = of->kn->priv;
@@ -365,14 +365,16 @@ int resctrl_group_mondata_show(struct seq_file *m, void *arg)
 
 	/* show monitor data */
 	d = mpam_find_domain(r, md.u.domid, NULL);
-	if (IS_ERR(d)) {
+	if (IS_ERR_OR_NULL(d)) {
 		pr_warn("Could't find domain id %d\n", md.u.domid);
-		return -ENOENT;
+		ret = -ENOENT;
+		goto out;
 	}
 
 	usage = rr->mon_read(d, rdtgrp);
 	seq_printf(m, "%llu\n", usage);
 
+out:
 	put_resource_name(resname);
 	resctrl_group_kn_unlock(of->kn);
 	return ret;
-- 
2.31.1

